packages/esm-amd-loader/src/esm-amd-loader.ts

```diff
@@ -18,42 +18,331 @@ interface Window {
+        // unique keys in the registry. We don't want to see that here.
+        url: (module.isTopLevel === true) ? baseUrl : module.url
+      });
+      continue;
+    }
+
+    // We have a dependency on a real module.
+    const dependency = getModule(resolveUrl(module.urlBase, depSpec));
+    args.push(dependency.exports);
+    depModules.push(dependency);
+
+    if (dependency.stateData.state === StateEnum.Initialized) {
+      load(dependency as Module<Initialized>);
+    }
+  }
+  return [args, depModules];
+}
+
+/**
+ * Start executing our dependencies, in order, as they become available.
+ * Once they're all executed, execute our own module body, if any.
+ */
+function beginWaitingOnDeps(module: Module<WaitingOnEarlierScripts>) {
+  const stateData: WaitingOnDeps = {
+    state: StateEnum.WaitingOnDeps,
+    args: module.stateData.args,
+    deps: module.stateData.deps,
+    moduleBody: module.stateData.moduleBody
+  };
+  const mutatedModule = stateTransition(module, stateData);
+  waitOnDeps(
+      module.stateData.deps,
+      () => execute(mutatedModule),
+      (e) => fail(mutatedModule, e));
+  return mutatedModule;
+}
+
+/** Runs the given module body. */
+function execute(module: Module<WaitingOnDeps>): Module<Executed|Failed> {
+  const stateData = module.stateData;
+  if (stateData.moduleBody != null) {
+    try {
+      stateData.moduleBody.apply(null, stateData.args);
+    } catch (e) {
+      return fail(module, e);
+    }
+  }
+  return stateTransition(module, {state: StateEnum.Executed});
+}
+
+/**
+ * Called when a module has failed to load, either becuase its script errored,
+ * or because one of its transitive dependencies errored.
+ */
+function fail(mod: Module, error: Error) {
+  return stateTransition(mod, {state: StateEnum.Failed, error});
+}
+
+/**
+ * Transition the given module to the given state.
+ *
+ * Does not do any checking that the transition is legal.
+ * Calls onNextStateChange callbacks.
+ */
+function stateTransition<SD extends ModuleStateData>(
+    module: Module, stateData: SD): Module<SD> {
+  if (debugging) {
+    console.log(`${module.url} transitioning to state ${stateData.state}`);
+  }
+  const mutatedModule = module as Module<SD>;
+  mutatedModule.stateData = stateData;
+  if (mutatedModule.onNextStateChange.length > 0) {
+    const callbacks = mutatedModule.onNextStateChange.slice();
+    mutatedModule.onNextStateChange.length = 0;
+    for (const callback of callbacks) {
+      callback();
+    }
+  }
+  return mutatedModule;
+}
+
+function waitOnDeps(
+    deps: Module[],
+    onResolve: ResolveCallback|undefined,
+    onError: ErrorCallback|undefined): void {
+  if (deps.length === 0) {
+    if (onResolve) {
+      onResolve();
+    }
+    return;
+  }
+  const nextDep = deps[0];
+  switch (nextDep.stateData.state) {
+    // Cases where the dep needs us to push it forward.
+    case StateEnum.Initialized:
+      load(nextDep as Module<Initialized>);
```

Comment:
this isn't clear to me that loads aren't serialized. Why only load the nextDep?

Reply:
Honestly I don't _think_ this line of code is ever executed, as `beginWaitingOnEarlierScripts` should have ensured that all of our dependencies have made it at least to `Loading`. The tests all pass with this line replaced with `throw new Error()`, this is just a belt and suspenders sort of check. Maybe unnecessary? WDYT